### PR TITLE
Fix unique timer multi instance

### PR DESCRIPTION
**Pull request by vingu-linaro**
Fixes a regression introduced with fork event.
The unique timer were no more per instance but shared between all instance

### CODE REVIEW

File: src/rt-app_parse_config.c [hunk 1]
```diff
@@ -325,40 +335,50 @@ static void
...
 	if (json_object_is_type(resources, json_type_object)) {
-		opts->nresources = 0;
+		opts->resources->nresources = 0;
 		foreach(resources, entry, key, val, idx) {
-			opts->nresources++;
+			opts->resources->nresources++;
 		}
 
-		log_info(PFX "Found %d Resources", opts->nresources);
-		opts->resources = malloc(sizeof(rtapp_resource_t) * opts->nresources);
```
Comment (qais-yousef) [inline src/rt-app_parse_config.c]: Sigh. I see the code already doesn't do that. Looking at the code we have a few of these actually. Oh well.

File: src/rt-app_parse_config.c [hunk 2]
```diff
@@ -325,40 +335,50 @@ static void
...
+		log_info(PFX "Found %d Resources", opts->resources->nresources);
+		size = sizeof(rtapp_resources_t) + sizeof(rtapp_resource_t) * opts->resources->nresources;
 
+		opts->resources = realloc(opts->resources, size);
```
Comment (qais-yousef) [inline src/rt-app_parse_config.c]: Validate the return value?

File: src/rt-app_parse_config.c [hunk 3]
```diff
@@ -830,14 +853,28 @@ parse_task_phase_data(struct json_object *obj,
...
 	struct json_object *phases_obj, *resources;
 
 	log_info(PFX "Parsing task %s [%d]", name, index);
 
 	/* common and defaults */
-	data->resources = &opts->resources;
+	data->opts = opts;
+
+	/*
+	 * Set a pointer to opt entry as we miht end up to modify the localtion of
```
Comment (qais-yousef) [inline src/rt-app_parse_config.c]: s/miht/might/

File: src/rt-app_parse_config.c [hunk 4]
```diff
@@ -830,14 +853,28 @@ parse_task_phase_data(struct json_object *obj,
...
+	 * global resources during realloc
+	 */
+	data->global_resources = &(opts->resources);
+
+	/*
+	 * Create at least a "empty" struct that will then be filled either will
+	 * parsing resources table or when parsing task's event
+	 */
+	data->local_resources = malloc(sizeof(rtapp_resources_t));
```
Comment (qais-yousef) [inline src/rt-app_parse_config.c]: Check returned value is valid

File: src/rt-app_types.h [hunk 5]
```diff
@@ -206,6 +212,10 @@ typedef struct _thread_data_t {
 
 	int forked;
 	int num_instances;
+
+	struct _rtapp_options_t *opts;
+	rtapp_resources_t *local_resources;
+	rtapp_resources_t **global_resources;
```
Comment (qais-yousef) [inline src/rt-app_types.h]: Does gloabal_resources need to be a double pointer? It seems to me an artefact left over from previous code?

File: src/rt-app_parse_config.c [hunk 6]
```diff
@@ -830,14 +853,28 @@ parse_task_phase_data(struct json_object *obj,
 
 static void
 parse_task_data(char *name, struct json_object *obj, int index,
-		  thread_data_t *data, rtapp_options_t *opts)
+			thread_data_t *data, rtapp_options_t *opts)
```
Comment (qais-yousef) [inline src/rt-app_parse_config.c]: Indentation change? It looks broken in this view at least - but if it looks OK in a proper text editor then ignore me.

File: src/rt-app.c [hunk 7]
```diff
@@ -108,6 +108,26 @@ static void thread_data_set_unique_name(thread_data_t *tdata, int nforks)
...
+/*
+ * Give each created thread a list of unique resources.
+ */
+static int thread_data_create_unique_resources(thread_data_t *tdata, const thread_data_t *td)
+{
+	rtapp_resources_t *table = td->local_resources;
+	int local_size = sizeof(rtapp_resources_t) + sizeof(rtapp_resource_t) * table->nresources;
+
+	tdata->local_resources = malloc(local_size);
+	//log_info(PIN2 "thread_data_create_unique_resources %x size %d", tdata->local_resources, size);
```
[the remark this answers is not in the document]
Comment (vingu-linaro) [reply src/rt-app.c]: yes. will remove

File: src/rt-app.c [hunk 8]
```diff
@@ -619,14 +648,20 @@ int run(int ind,
...
-	for (i = 0; i <  opts.nresources; i++) {
-		if (opts.resources[i].type == rtapp_wait) {
-			pthread_cond_broadcast(&opts.resources[i].res.cond.obj);
+	/*
+	 * Force wake up of all waiting threads.
+	 * At now we don't need to look into local resources because rtapp_wait
+	 * and rtapp_barrier are always global
+	 */
+	for (i = 0; i <  nresources; i++) {
+		if (resources[i].type == rtapp_wait) {
```
[the remark this answers is not in the document]
Comment (vingu-linaro) [reply src/rt-app.c]: maybe but let consider that in a dedicated clean up pul request

File: src/rt-app_parse_config.c [hunk 9]
```diff
@@ -301,22 +303,30 @@ parse_resource_data(const char *name, struct json_object *obj, int idx,
...
 	log_info(PIN2 "Add new resource %s [%d] type %d", name, idx, type);
 
-	opts->nresources++;
-	opts->resources = realloc(opts->resources, sizeof(rtapp_resource_t) * opts->nresources);
+	table->nresources++;
 
-	init_resource_data(name, type, idx, opts);
+	size = sizeof(rtapp_resources_t) + sizeof(rtapp_resource_t) * table->nresources;
+
+	*resources_table = realloc(*resources_table, size);
```
[the remark this answers is not in the document]
Comment (vingu-linaro) [reply src/rt-app_parse_config.c]: yes. I have to add the error path
more generally speaking, we need to check all error path

File: src/rt-app_parse_config.c [hunk 10]
```diff
@@ -325,40 +335,50 @@ static void
...
 	struct lh_entry *entry; char *key; struct json_object *val; int idx;
+	int size;
 
 	log_info(PFX "Parsing resource section");
 
+	/*
+	 * Create at least a "empty" struct that will then be filled either will
+	 * parsing resources table or when parsing task's event
+	 */
+	opts->resources = malloc(sizeof(rtapp_resources_t));
```
[the remark this answers is not in the document]
Comment (vingu-linaro) [reply src/rt-app_parse_config.c]: yes

File: src/rt-app_parse_config.c (hunk 6 above)
Comment (vingu-linaro) [reply src/rt-app_parse_config.c]: I will check

File: src/rt-app_parse_config.c (hunk 3 above)
Comment (vingu-linaro) [reply src/rt-app_parse_config.c]: yes

File: src/rt-app_types.h (hunk 5 above)
Comment (vingu-linaro) [reply src/rt-app_types.h]: yes it has to be double pointer
because realloc can modify the pointer, we must be able to save the new pointer.

Comment (qais-yousef) [reply src/rt-app_types.h]: Maybe it's friday afternoon that's clouding my thinking. But we can save the returned pointer with a single pointer too. What am I missing? When forking we use realloc to expand `threads` structure - which is a single pointer. If we fail to expand then we basically fail to run the whole test and bail out.

Comment (vingu-linaro) [reply src/rt-app_types.h]: realloc can move the  struct somewhere else so we must update the place where the pointer is saved. In fact, this happened during some tests.
If you look at add_resource_data(), we don't use the local pointer table but *resources_table after the realloc

Comment (qais-yousef) [reply src/rt-app_types.h]: >If you look at add_resource_data(), we don't use the local pointer table but *resources_table after the realloc

Yes because it's a double pointer. If you make it a single pointer you can just access it directly, no?

Because you added rtapp_resources_t there's no need for the double pointer AFAICS and I don't see why local_resources can be a single pointer but not global_resources since the table is now embedded in the new struct.

We expand threads in fork event in rt-app.c using realloc, why the usage of realloc in the 2 cases is not equivalent?

Comment (vingu-linaro) [reply src/rt-app_types.h]: no, otherwise we would only update the "local" pointer of gloabl_resources of the task that has added the resources but other tasks will continue to point to previous one.
single pointer is enough for loacl_resources because it's not point by others.

we have opts.resources->the_global_resouces and threadA_data.global_resources-> opts.resources->the_global_resouces and threadB_data.global_resources-> opts.resources->the_global_resouces. If threadA_data and threadB_data were directly pointing to the_global_resouces instead, if @ of the_global_resouces changes, only the thread that did the changes would have the new @ others would still point to previous @

Comment (qais-yousef) [reply src/rt-app_types.h]: Oh I see what I was missing now. This is a per thread pointer to a global resource. So we are saving a pointer to a pointer actually - I was thinking the table stored directly in this struct and missed it's part of the tread data, my bad.

If we already have a pointer to the global_resources in opts then we should just use that IMHO. It's very easy to create a micro to improve readability. But I don't feel strongly about it.

Comment (vingu-linaro) [reply src/rt-app_types.h]: This pointer is used quite often when running events and make the code more readable instead of finding our way to the pointer every time.

Comment (vingu-linaro) [reply src/rt-app_types.h]: opts pointer is now removed from thread structure because it was only used for resources init like pi_enable

**Comment by vingu-linaro** [comment]
I have updated the fix and remove the finally useless opts pointer.
I have also checked the returned value of meme allocation

**Comment by vingu-linaro** [comment]
Have you got comments ? otherwise i'm going to merge it

**Comment by qais-yousef** [comment]
The change looks fine to me. I don't know if @jlelli wants to add anything.

It would have been nice to split the error checking cleanups into a separate patch or update the commit message to mention this cleanup at least.

File: src/rt-app.c [hunk 11]
```diff
@@ -495,18 +525,18 @@ static int run_event(event_data_t *event, int dry_run,
 		break;
 	case rtapp_suspend:
 		{
-		log_debug("suspend %s ", rdata->name);
-		pthread_mutex_lock(&(ddata->res.mtx.obj));
-		pthread_cond_wait(&(rdata->res.cond.obj), &(ddata->res.mtx.obj));
-		pthread_mutex_unlock(&(ddata->res.mtx.obj));
+			log_debug("suspend %s ", rdata->name);
```
Comment (jlelli) [inline src/rt-app.c]: This and right below seems unrelated to this change?

File: src/rt-app.c [hunk 12]
```diff
@@ -712,6 +748,11 @@ static void create_cpuset_str(cpuset_data_t *cpu_data)
 	}
 
 	cpu_data->cpuset_str = malloc(size_needed);
+	if (!cpu_data->cpuset_str) {
```
Comment (jlelli) [inline src/rt-app.c]: This unrelated as well?

File: src/rt-app.c [hunk 13]
```diff
@@ -952,6 +995,10 @@ void *thread_body(void *arg)
 	/* Init timing buffer */
 	if (opts.logsize > 0) {
 		timings = malloc(opts.logsize);
+		/*
```
Comment (jlelli) [inline src/rt-app.c]: Unrelated?

File: src/rt-app.c [hunk 14]
```diff
@@ -1177,6 +1223,10 @@ int main(int argc, char* argv[])
 	/* allocated threads */
 	nthreads = opts.nthreads;
 	threads = malloc(nthreads * sizeof(pthread_t));
+	if (!threads) {
```
Comment (jlelli) [inline src/rt-app.c]: Unrelated?

File: src/rt-app_parse_config.c [hunk 15]
```diff
@@ -526,27 +561,27 @@ parse_task_event_data(char *name, struct json_object *obj,
 	if (!strncmp(name, "timer", strlen("timer"))) {
 
 		tmp = get_string_value_from(obj, "ref", TRUE, "unknown");
-		if (!strncmp(tmp, "unique", strlen("unique")))
-				ref = create_unique_name(unique_name, sizeof(unique_name), tmp, tag);
-		else
-				ref = tmp;
 
-		i = get_resource_index(ref, rtapp_timer, opts);
+		if (!strncmp(tmp, "unique", strlen("unique"))) {
```
Comment (jlelli) [inline src/rt-app_parse_config.c]: Mmm. I'm wondering if for forking tasks timers should be unique by default?

File: src/rt-app_parse_config.c [hunk 16]
```diff
@@ -526,27 +561,27 @@ parse_task_event_data(char *name, struct json_object *obj,
...
+			ref = create_unique_name(unique_name, sizeof(unique_name), tmp, tag);
+			resources_table = &tdata->local_resources;
+			data->type = rtapp_timer_unique;
```
Comment (jlelli) [inline src/rt-app_parse_config.c]: Do we need a different type? Can't we make timers local even for tasks that don't fork?

File: src/rt-app.c (hunk 11 above)
Comment (vingu-linaro) [reply src/rt-app.c]: the 1st commit is doing a cleanup 
the 2nd fix the unique timer problem

File: src/rt-app.c (hunk 12 above)
Comment (vingu-linaro) [reply src/rt-app.c]: While fixing unique timer, missing error path has been highlighted. All this are in the 1st patch

File: src/rt-app_parse_config.c (hunk 15 above)
Comment (vingu-linaro) [reply src/rt-app_parse_config.c]: I don't think. I have in mind some UC where threads fork but still use the same timer

File: src/rt-app_parse_config.c (hunk 16 above)
Comment (vingu-linaro) [reply src/rt-app_parse_config.c]: not sure to get your point. "unique" is already a key of rt-app timer description
Any timer with a "unique" name will be unique/local to each thread whatever the thread is created at the beg or forked later.

File: src/rt-app_parse_config.c (hunk 15 above)
Comment (jlelli) [reply src/rt-app_parse_config.c]: ok

File: src/rt-app_parse_config.c (hunk 16 above)
Comment (jlelli) [reply src/rt-app_parse_config.c]: You actually answered my doubt in question above.. there are cases where tasks need to share timers.. so we need that distinction.

**Comment by vingu-linaro** [comment]
I have updated patchset:
-  No code change
- move some cleanup changes that were in patch2 into patch1.

**Comment by jlelli** [comment]
LGTM, tested as well.
I'm going to merge this, thanks a lot @vingu-linaro .
One last thing that I was thinking however is that it would be nice to document how we handle resources (data structures, etc.). Do you think you could add some of that somewhere as a comment?

**Comment by vingu-linaro** [comment]
I'm going to prepare a patch to add such details

**Comment by jlelli** [comment]
Awesome!
